rustdoc: Foldable impl blocks

vi opened the pull request:
Addresses #40363, #45720, #24483, #23986 and so on

* Expands and refactors collapseDocs and toggleAllDocs
* Adds [-] toggle to all impls (including inherent impl)
* Makes it hiding though main css file, not though element inline style

May need to be addressed:

* "[-]" and anchor link copier are overlaid a bit
* Inherent methods are also hidden by the global [-] toggle.
* Auto-collapsing "Iterator" and so on by default is not implemented yet
* Tested only shallowly and only in Chromiuim
* No tests. Are there tests for css/js part here?
* The new implementation may be a bit slower.

What next steps are need to be done before the integration?

rust-highfive commented:
Some changes occurred in HTML/CSS.

cc @GuillaumeGomez

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @GuillaumeGomez (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

kennytm commented:
Thanks for the PR! We’ll periodically check in on it to make sure that @GuillaumeGomez   or someone else from the team reviews it soon.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+                }
+            }
+
+            if(action == "show") {
```
Missing whitespace after `if`.

GuillaumeGomez replied:
Also, please use `===`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+
+            var action = mode;
+            if (action == "toggle") {
+                if(hasClass(relatedDoc, "fns-now-collapsed")) {
```
Missing whitespace after `if`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+                relatedDoc = relatedDoc.nextElementSibling;
+            }
+
+            if (!relatedDoc) return;
```
Please don't do that. Use `{}`, it makes reading easier.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+                            removeClass(n, "hidden-by-impl-hider");
+                        }
+                        var ns = n.nextElementSibling;
+                        while(true) {
```
Missing whitespace after `while`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+            // we are collapsing the impl block
+            function implHider(addOrRemove) {
+                return function(n) {
+                    if(hasClass(n, "method")) {
```
Missing whitespace after `if`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+            if (hasClass(relatedDoc, "docblock")) {
+                var action = mode;
+                if (action == "toggle") {
+                    if(hasClass(relatedDoc, "hidden-by-usual-hider")) {
```
Missing whitespace after `if`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+            var action = mode;
+            if (action == "toggle") {
+                if(hasClass(relatedDoc, "fns-now-collapsed")) {
+                    action="show";
```
Missing whitespace around `=`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+                if(hasClass(relatedDoc, "fns-now-collapsed")) {
+                    action="show";
+                } else {
+                    action="hide";
```
Missing whitespace around `=`.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,74 +1618,124 @@
+                removeClass(relatedDoc, "fns-now-collapsed");
+                onEach(toggle.childNodes, adjustToggle(false));
+                onEach(relatedDoc.childNodes, implHider(false));
+            } else if (action == "hide") {
```
Please use `===`.

GuillaumeGomez commented on src/librustdoc/html/static/themes/dark.css:
```diff
@@ -386,3 +386,8 @@ kbd {
 	}
 }
+
+.hidden-by-impl-hider, 
```
This rule has nothing to do in themes. Please put in `rustdoc.css` at the upper level.

GuillaumeGomez commented on src/librustdoc/html/static/themes/main.css:
```diff
@@ -383,3 +383,8 @@ kbd {
 	}
 }
+
+.hidden-by-impl-hider, 
```
This rule has nothing to do in themes. Please put in `rustdoc.css` at the upper level.

GuillaumeGomez commented:
I made a first quick review. I'll test what it's doing this evening or tomorrow. Seems globally good otherwise (just nits mainly) .

vi commented:
Is there maybe some automatic script that makes whitespace in javascript files meet the standard in one move?

vi commented:
Where the list of collapsed-by-default traits should reside? In Rust part (marking html elements with a class like "collapsed-by-default") or in Javascript part?

Thomasdezeeuw commented:
@vi Why not collapse all trait implementations by default? In most cases you want to visit the actual trait documentation anyway, since the documentation at the implementation only shows a single line or so per method.

vi commented:
A good idea. Important things like associated types/consts or impl-level documentation are visible even when collapsed anyway.

vi commented:
How do I address this overlaying:

![screenshot](https://user-images.githubusercontent.com/173219/35631694-9421b9f4-06b5-11e8-9d56-f492d77c375a.png)

Also why don't inherent methods have such anchored link copier?

kennytm commented:
Could we just make the `impl` keyword function as that § anchor (and remove the §)? Too implicit?

vi commented:
Without the § it won't be obvious that anchor links are supported in general.

GuillaumeGomez commented:
Normally, `[-]`/`[+]` should be display block. If not, just increase the `margin` or the `right`/`left` value (depending which is used in this case, don't remember...).

vi commented:
Is moving those handles to the left a good idea?
![screenshot](https://user-images.githubusercontent.com/173219/35681536-bf5d6cf8-076e-11e8-9710-e3773c259ff5.png)


```diff
diff --git a/rustdoc.css b/rustdoc.css
index d4b9fba..d92685b 100644
--- a/rustdoc.css
+++ b/rustdoc.css
@@ -695,7 +695,7 @@ a.test-arrow:hover{
 .collapse-toggle {
        font-weight: 300;
        position: absolute;
-       left: -23px;
+       left: -40px;
        top: 0;
 }
```

It also reseves space for further § for function links (which do not appear now).

GuillaumeGomez commented:
No because the `i` thing is already there.

GuillaumeGomez commented:
And you have to take into account the mobile version as well.

vi commented:
> No because the i thing is already there.

Meaning that "important traits" popup?

> And you have to take into account the mobile version as well.

When I activate "mobile view" in Chromium, collapse toggles for impls get outside viewport...

How do I get collapse toggles, §-s and circled i-s all available, non-intersecting and not taking too much space when not in use? Will this list be extened even more in future?

GuillaumeGomez commented:
It wasn't supposed to get extended this quickly. If that's an issue in your case, we'll need to think about another way to have access to those pieces of information.

vi commented:
Maybe the circled "i" should be at right (also closer to return type than to function name), not at the left?

GuillaumeGomez commented:
The "i" can't be at right. It'd reduce the potential size of the functions' name. Also, on the right, there is already `[src]` and (in some cases) version numbers. The right side is crowded. I'm starting to think that we should maybe group buttons when there are more than one (if there is "i" and your new button at the same time, we group them, otherwise they're just there).

This design issue is quite frustrating haha.

vi commented:
@GuillaumeGomez , Added `!important` so folding works on small screens as well.

Reverted moving the toggler to left, so it is should be usable on mobile. It looks not nice and intersects with the §...

Here is preview in mobile mode:

![screenshot](https://user-images.githubusercontent.com/173219/35886052-d614e094-0ba0-11e8-8d1a-776e4926f38f.png)

So what to do next about it?

GuillaumeGomez commented:
Aren't the `[-]` and `[+]` closer to the text now? Otherwise it seems globally good. I don't have much to say against it. I'll need to confirm for desktop version too.

I'd also like to have @QuietMisdreavus's confirmation. :)

QuietMisdreavus commented:
Immediate reaction: i'm a little skittish about folding all traits, but i think i can get used to it. There are a few cases where a trait is really the primary interaction with a given type (e.g. the traits that currently get put in the "Important Traits" popup) and folding those up feels awkward, especially if the type has no other methods. However, even then, just seeing `impl Iterator for Something` is possibly all the information you need, just like any other trait, so it's fine.

Now, onto the presentation. I'm assuming @GuillaumeGomez is fine with the JS, so i won't dive too far into the implementation.

One of the first things i noticed was that the only things included in the fold were methods. Anything else, like associated items or docs on the impl block itself, were not included. Was this deliberate? It seems odd that only some of the impl is being hidden. It also contrasts with "impls on foreign items", which does hide everything but the header.

![image](https://user-images.githubusercontent.com/5217170/35942639-ef52096c-0c1b-11e8-94ad-89d4cd482abc.png)

![image](https://user-images.githubusercontent.com/5217170/35943629-451169bc-0c1f-11e8-9b8c-bdc6caeb4d9a.png)

As for the link anchor icon: Would it be possible to restrict the CSS change to only those link anchors on impl blocks? Those will never get the "important traits" icon right now, and are unlikely to receive it in the future.

vi commented:
> especially if the type has no other methods

Add a special rule that if there is only one impl and no inherent methods then don't fold the impl? I doubt it is that useful. Such types may just start disappearing once `impl Trait` come.

> One of the first things i noticed was that the only things included in the fold were methods. Anything else, like associated items or docs on the impl block itself, were not included. Was this deliberate

Yes. Methods are always the same for given trait, but associated types/constants differ from impl to impl. For example, for `Iterator` the most important thing is the `Item`. `impl`-level doc is rather rare thing, so it worth show it by default.

vi commented:
> It also contrasts with "impls on foreign items", which does hide everything but the header.

If associated types / constants of impls on foreigh types are hidden then it is probably a bug...

> Would it be possible to restrict the CSS change to only those link anchors on impl blocks?

Don't understand... Do you propose to make a change to CSS or to partially rollback ("restrict") already existing change? Currently the only change to CSS is the main `display: none` hider, pictures with collapse toggle moved to the left are outdated.

QuietMisdreavus commented:
> Yes. Methods are always the same for given trait, but associated types/constants differ from impl to impl.

Ah, this makes sense. As for the "implementations on foreign types" section, i would agree with you (that it's a bug to fold everything), given this rationale. Currently it wraps everything in a `.docblock` when emitting the section, but that can probably be changed.

> Don't understand... Do you propose to make a change to CSS or to partially rollback ("restrict") already existing change? Currently the only change to CSS is the main display: none hider, pictures with collapse toggle moved to the left are outdated.

I was referring to the rejected CSS change discussed in https://github.com/rust-lang/rust/pull/47894#issuecomment-362269114 where you wanted to move the "fold" button to make room for the link anchor. But maybe i'm not fully understanding the reason for its rejection. `>_>`

vi commented:
> But maybe i'm not fully understanding the reason for its rejection.

The main reason is that on small (mobile) screens it gets out from viewport entirely. (Without moving it is only slightly obscured by the edge of screen which can be seen in screenshot about, but still clickable).

pietroalbini commented:
@QuietMisdreavus what's the status on this PR?

QuietMisdreavus commented:
I just discussed this with @GuillaumeGomez, and he said he wants to figure out some way to stop this overlap from happening:

![image](https://user-images.githubusercontent.com/5217170/36105965-94a4ccc0-0fdb-11e8-8f38-59f7e683439c.png)

Until then, we're holding off from merging this PR.

GuillaumeGomez commented on src/librustdoc/html/static/main.js:
```diff
@@ -1618,76 +1618,143 @@
+                            if (ns && (
+                                    hasClass(ns, "docblock") ||
+                                    hasClass(ns, "stability") ||
+                                    false
```
Why this line?

vi replied:
To keep the list open and allow 1-line diffs for removals and additions of classes. To avlid the first or last mentioned class being special (by not having `||` somewhere).

GuillaumeGomez commented:
So there is still the issue about the `§` (when you hover). To fix this, I suggest adding a padding on the left of impl blocks so the `[+]` is directly on the left and the `§` can appear without overlap. What do you think about it @vi?

vi commented:
Maybe. There is also one stated issue about over-hiding the foreign impl blocks.

QuietMisdreavus commented:
Over-hiding foreign impl blocks is unrelated to this PR. That can happen separately (and independently, tbh).

shepmaster commented:
> I suggest adding a padding

It looks like this is back in your court, @vi !

vi commented:
Done,

![screenshot1](https://user-images.githubusercontent.com/173219/36343920-bed8fa10-1423-11e8-8978-bb0a75d57232.png)

![screenshot](https://user-images.githubusercontent.com/173219/36343929-e7fa3de6-1423-11e8-846e-2e76d3231a75.png)

GuillaumeGomez commented:
A bit more spacing for "§" please! :D

vi commented:
@GuillaumeGomez, On desktop version (first screenshot) or on mobile version (second screenshot)?

On the desktop I don't think more spacing is needed. 

Maybe better to just omit § entirely on mobile version, to economize screen space?

GuillaumeGomez commented:
On desktop. It's useless for mobile since you can't hover a method properly speaking.

vi commented:
@GuillaumeGomez , Then annotate or edit [the last screenshot](https://user-images.githubusercontent.com/173219/36343920-bed8fa10-1423-11e8-8978-bb0a75d57232.png) with a positions of [+] and § that you prefer.

GuillaumeGomez commented:
Add 2 pixels to the right of "§".

vi commented:
Done:

![screenshot](https://user-images.githubusercontent.com/173219/36395766-d6a4948e-15cb-11e8-9004-d416109da558.png)

GuillaumeGomez commented on src/librustdoc/html/static/rustdoc.css:
```diff
@@ -906,6 +906,10 @@ span.since {
 		display: flex;
 	}
+
+	.anchor {
```
There isn't supposed to have an anchor in mobile version normally.

vi replied:
But it appears when I click at the impl in Chromium's mobile view.

It can be removed with  `display:none !important`, then the `[+]` may be moved further right.

GuillaumeGomez replied:
I approve (for the mobile version)!

GuillaumeGomez commented:
Ok, all good for me. Anything else for you @QuietMisdreavus?

QuietMisdreavus commented:
Looks good to me!

@bors r=GuillaumeGomez,QuietMisdreavus

bors commented:
:pushpin: Commit df1b9a8 has been approved by `GuillaumeGomez,QuietMisdreavus`

<!-- @bors r=GuillaumeGomez,QuietMisdreavus df1b9a8584bf273fb807bae2d9dd55833aee5964 -->

bors commented:
:hourglass: Testing commit df1b9a8584bf273fb807bae2d9dd55833aee5964 with merge f7602112603dfa65f9d50cc4217d08b0162e4b8b...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6441)

kennytm commented:
@bors retry #46903

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/345651157?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry 

3 hour timeout in `dist-{powerpc,powerpc64le}-linux`.

bors commented:
:hourglass: Testing commit df1b9a8584bf273fb807bae2d9dd55833aee5964 with merge 2452e2bfc1b92af2cb68e162a3cad0f9d473f177...

bors commented:
:broken_heart: Test failed - [status-travis](https://travis-ci.org/rust-lang/rust/builds/346158035?utm_source=github_status&utm_medium=notification)

kennytm commented:
@bors retry 

3 hour timeout in `dist-s390x-linux`.

<details><summary>Timing breakdown</summary>

```
make-prepare                	  47.14
pytest/bootstrap            	   0.00
stage0-std                  	  63.00
stage0-test                 	  17.40
stage0-rustc                	1118.22
llvm                        	 302.07
stage0-trans-llvm           	 166.90
stage1-std                  	  62.31
stage1-test                 	  17.31
stage1-rustc                	1173.68
stage1-trans-llvm           	 181.68
stage1-std                  	  67.91
stage0-unstable-book-gen    	  10.28
stage0-rustbook             	 167.91
stage2-rustdoc              	 165.16
doc/std                     	  38.67
doc/compiler                	  16.36
stage1-rustc                	1177.74
doc/proc_macro              	  27.52
stage2-error_index_generator	   5.34
stage0-fabricate            	  90.44
llvm                        	2465.59
stage1-trans-llvm           	 189.50
stage2-rustdoc              	 160.93
stage2-cargo                	 401.38
stage2-rustfmt              	 240.86
stage2-cargo-fmt            	   0.00
```

</details>

41 minutes spent on compiling LLVM (for s390x), maybe sccache network error?

bors commented:
:hourglass: Testing commit df1b9a8584bf273fb807bae2d9dd55833aee5964 with merge 89e5a0796e0c5de52cd1d6c7541e5c2c7c47f2cd...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6490), [status-travis](https://travis-ci.org/rust-lang/rust/builds/347116005?utm_source=github_status&utm_medium=notification)
Approved by: GuillaumeGomez,QuietMisdreavus
Pushing 89e5a0796e0c5de52cd1d6c7541e5c2c7c47f2cd to master...

vi commented:
Looks like § started to intersect unrelated headings (at least on docs.rs):

![screenshot](https://user-images.githubusercontent.com/173219/37494309-4fb07dc0-28b9-11e8-965d-3ecbf1946805.png)

Maybe a follow-up pull request is needed. Or shall entire § handling be overhauled? For example, it does not allow to get an anchored link to a particular function inside impl.

shepmaster commented:
> Or shall entire § handling be overhauled?

I often find the § disappearing as I move the cursor from the header to the § in order to copy the link, so I'd be in favor of fixing that ;-).

GuillaumeGomez commented:
It's been fixed in rustdoc already.
